File: Core/Company-Interview/GoldmanSach/FractionToRecurringDecimal/FractionToRecurringDecimal.java
package GoldmanSachs.FractionToRecurringDecimal;

import java.util.HashMap;
import java.util.Map;


public class FractionToRecurringDecimal {

    public static void main(String[] args) {
        FractionToRecurringDecimal fracToDec = new FractionToRecurringDecimal();
        System.out.println(fracToDec.fractionToDecimal(-1, 3));   // -0.(3)
        System.out.println(fracToDec.fractionToDecimal(1, 3));    // 0.(3)
        System.out.println(fracToDec.fractionToDecimal(-1, -2));  // 0.5
        System.out.println(fracToDec.fractionToDecimal(1, 4));    // 0.25
        System.out.println(fracToDec.fractionToDecimal(0, -2)); // 0
        System.out.println(fracToDec.fractionToDecimal(Integer.MIN_VALUE, 1));
    }

    /**
     * Converts a fraction to its decimal representation, detecting repeating patterns.
     * Time Complexity: O(Denominator) - In the worst case, we can only have as many unique
     * remainders as the value of the denominator before a cycle occurs.
     * Space Complexity: O(Denominator) - To store the position of each remainder in our map.
     */
    public String fractionToDecimal(int numerator, int denominator) {
        // 1. Validation & Edge Cases
        if (denominator == 0) throw new ArithmeticException("cannot divide by zero");
        if (numerator == 0) return "0";

        StringBuilder result = new StringBuilder();

        // 2. Sign Handling:
        // Use XOR (^) to determine if the result is negative.
        // This handles the case where one is negative and the other is positive.
        if ((numerator < 0) ^ (denominator < 0)) {
            result.append("-");
        }

        // Convert to long to prevent overflow (e.g., Integer.MIN_VALUE / -1)
        long num = Math.abs((long) numerator);
        long denom = Math.abs((long) denominator);

        // 3. Integer Part:
        // Append the whole number part and calculate the initial remainder.
        result.append(num / denom);
        long remainder = num % denom;

        // If there is no remainder, the division is exact.
        if (remainder == 0) return result.toString();

        // 4. Fractional Part:
        result.append(".");

        /* * Map tracks <Remainder, Position in StringBuilder>.
        * If we see the same remainder again, it means the sequence of digits
        * from that previous position will now repeat infinitely.
        */
        Map<Long, Integer> remainderMap = new HashMap<>();

        while (remainder != 0) {
            // Cycle Detection:
            if (remainderMap.containsKey(remainder)) {
                // We retrieve the index where we first saw this remainder.
                // This is exactly where the repeating digits started being appended.
                int cycleStart = remainderMap.get(remainder);
                result.insert(cycleStart, "("); // Wrap the start of the repeating part
                result.append(")");             // End the repeating part
                return result.toString();
            }

            // 3. THE "SNAPSHOT" (Recording):
            // Before we calculate the next digit, we record the current remainder
            // and the current length of our StringBuilder.
            // This length acts as a 'bookmark' for where the digit resulting
            // from this specific remainder will be placed.

            // Store current remainder and the index where its resulting digit will go
            remainderMap.put(remainder, result.length());

            // Standard Long Division:
            // Multiply remainder by 10 to "bring down a zero"
            remainder *= 10;
            result.append(remainder / denom);
            remainder %= denom;
        }

        return result.toString();
    }
}

File: Core/Algorithms/FractionToRecurringDecimal/fraction-to-recurring-decimal.py
"""
Time Complexity
Space Complexity
"""


def fractionToDecimalBrute(numerator: int, denominator: int) -> str:
    pass


"""
Time Complexity: O(n) - at most n unique remainders before a repeat
Space Complexity: O(n) -  dictionary holds at most n entries
"""


def fractionToDecimalTest(num: int, denom: int) -> str:
    # denom cannot be zero
    if denom == 0:
        raise ZeroDivisionError("cannot divide by zero")

    result = []
    remainder_map = {}

    # Edge case: zero numerator
    if num == 0:
        return "0"

    # --- Step 1: Handle sign ---
    # XOR: negative only if exactly one is negative
    # with OR, any negative input triggers the "-" sign so "-1/-2" will give "-0.5"
    if (num < 0) ^ (denom < 0):
        result.append("-")

    # case when both are negative
    num, denom = abs(num), abs(denom)

    # Step 2: Handle Integer part
    result.append(str(num // denom))
    remainder = num % denom

    # No decimal part needed
    if remainder == 0:
        return "".join(result)

    # Step 3: Handle decimal
    result.append(".")

    while remainder:
        # Cycle detected → insert parentheses at the stored index
        if remainder in remainder_map:
            # do something
            cycle_start = remainder_map[remainder]
            result.insert(cycle_start, "(")
            result.append(")")
            return "".join(result)

        # Record current position before appending the digit
        remainder_map[remainder] = len(result)

        # Long division step
        remainder *= 10
        result.append(str(remainder // denom))
        remainder %= denom

    return "".join(result)


# print(fractionToDecimal(-1, 0))
# print(fractionToDecimal(0, -2))
# print(fractionToDecimal(-2, -1))
# print(fractionToDecimal(-1, -2))
# print(fractionToDecimal(1, 3))
# print(fractionToDecimal(-1, -2))
# print(fractionToDecimal(-1, -8))


# TC: O(n), SC: O(n)
def fractionToDecimalTest(num: int, denom: int) -> str:
    """
    case 0: when denom is 0
    case 1: when num is 0
    # handle sign
    case 2: when one of num or demon is -ve
    case 3: when both num/denom are -ve
    # handle digit
    case 4: when there is no remainder
    # handle decimal
    case 5: when there is a remainder
    """

    result = []
    remainder_map = {}

    if not isinstance(num, int) or not isinstance(denom, int):
        return None

    if denom == 0:
        raise ZeroDivisionError("cannot divide by zero")

    if num == 0:
        return "0"

    # handle sign
    # when exactly one of the values are -ve
    if (num < 0) ^ (denom < 0):
        result.append("-")

    # here both are -ve
    num, denom = abs(num), abs(denom)

    # handle digits
    quotient = num // denom
    remainder = num % denom
    result.append(str(quotient))

    if remainder == 0:
        return "".join(result)

    # handle decimal (where there is a remainder)
    result.append(".")

    while remainder:
        if remainder in remainder_map:
            cycle_start = remainder_map[remainder]
            result.insert(cycle_start, "(")
            result.append(")")
            return "".join(result)

        remainder_map[remainder] = len(result)

        # long division
        remainder *= 10
        quotient = remainder // denom
        result.append(str(quotient))
        remainder %= denom

    return "".join(result)


print(fractionToDecimalTest(-1, 3))
